mlflow/projects/__init__.py

```diff
@@ -37,28 +37,51 @@ class ExecutionException(Exception):
 
 def _run(uri, entry_point="main", version=None, parameters=None, experiment_id=None,
          mode=None, cluster_spec=None, git_username=None, git_password=None, use_conda=True,
-         use_temp_cwd=False, storage_dir=None, block=True):
+         use_temp_cwd=False, storage_dir=None, block=True, run_id=None):
+    """
+    Helper that delegates to the project-running method corresponding to the passed-in mode.
+    Returns a `SubmittedRun` corresponding to the project run.
+    """
     exp_id = experiment_id or tracking._get_experiment_id()
-    if mode is None or mode == "local":
-        return _run_local(
-            uri=uri, entry_point=entry_point, version=version, parameters=parameters,
-            experiment_id=exp_id, use_conda=use_conda, use_temp_cwd=use_temp_cwd,
-            storage_dir=storage_dir, git_username=git_username, git_password=git_password,
-            block=block)
     if mode == "databricks":
         from mlflow.projects.databricks import run_databricks
         return run_databricks(
             uri=uri, entry_point=entry_point, version=version, parameters=parameters,
             experiment_id=exp_id, cluster_spec=cluster_spec, git_username=git_username,
             git_password=git_password)
+    elif mode == "local" or mode is None:
+        # Fetch project into a working directory
+        work_dir = _fetch_project(uri, use_temp_cwd, version, git_username, git_password)
+        project = _load_project(project_dir=work_dir)
+        # Validate we specified correct params for the project
+        project.get_entry_point(entry_point)._validate_parameters(parameters)
+        if use_conda:
```

Comment:
This one might be worth a comment to mention that we have to load the conda env in the blocking section in order to avoid concurrency problems.